internal/sample/sample.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sample
+
+import (
+	"fmt"
+	"github.com/GoogleCloudPlatform/serverless-sample-tester/internal/gcloud"
+	"github.com/GoogleCloudPlatform/serverless-sample-tester/internal/lifecycle"
+	"github.com/GoogleCloudPlatform/serverless-sample-tester/internal/util"
+	"os/exec"
+	"strings"
+	"unicode"
+)
+
+//
+const maxCloudContainerImageTagLen = 53
+
+// Sample represents a Google Cloud Platform sample and associated properties.
+type Sample struct {
+	Name string
+
+	// The local directory this sample is located in.
+	Dir string
+
+	// The cloudRunService this sample will deploy to.
+	Service gcloud.CloudRunService
+
+	// The lifecycle for building and deploying this sample to Cloud Run.
+	BuildDeployLifecycle lifecycle.Lifecycle
+
+	// The URL location of this sample's build container image in the GCP Container Registry.
+	cloudContainerImageURL string
+}
+
+// NewSample creates a new sample object for the sample located in the provided local directory.
+func NewSample(dir string) (*Sample, error) {
+	name := sampleName(dir)
+
+	containerTag, err := cloudContainerImageTag(name)
+	if err != nil {
+		return nil, err
+	}
+
+	projectID, err := util.ExecCommand(util.GcloudCommandBuild("config", "get-value", "core/project"))
+	if err != nil {
+		return nil, err
+	}
+	cloudContainerImageURL := fmt.Sprintf("gcr.io/%s/%s", projectID, containerTag)
+
+	serviceName, err := gcloud.ServiceName(name)
+	if err != nil {
+		return nil, err
+	}
+	service := gcloud.CloudRunService{Name: serviceName}
+
+	buildDeployLifecycle := lifecycle.GetLifecycle(dir, service.Name, cloudContainerImageURL)
+
+	return &Sample{
+		Name:                   name,
+		Dir:                    dir,
+		Service:                service,
+		BuildDeployLifecycle:   buildDeployLifecycle,
+		cloudContainerImageURL: cloudContainerImageURL,
+	}, nil
```

Comment:
nit: it's probably good to separate the init and return values here for readability - otherwise the nil is a somewhat easy to miss.